### ISSUE TITLE
Global Options

### ISSUE THREAD

**Issue by tlovett** [useblocks/sphinxcontrib-needs#57]
Sometimes a useful attribute about a need can be derived rather than explicitly stated in the RST.

In our use case, we have a separate project that implements scripts that validate the requirement against some vendors implementation of the spec.  We'd like to keep track of which requirements have validation scripts written, but we'd also prefer not to have to keep track of this information in both the requirements project and the validation project.  The validation project already tracks traceability to the requirements in an usable data format such as YAML.   

It would be nice if we could add derived metadata to a requirement that a script exists and provide a link to that particular script.  

The idea I thought would be useful would be being able to specify dynamic attributes in the `conf.py` file, and map them to some callable that could derive the value at build time. 

Perhaps something like this in conf.py:

```Python
# Instantiate some object that can look up the information.
# In a more simple example, it could map to a single function

repos = ValidationRepository('some config to get the data')

needs_dynamic_options = {
   'validated' : repos.is_validated,
   'validation_link': repos.get_link,
}
``` 

These callables could be invoked and be passed the app environment and the need_info during the build, and then derive additional attributes.  This derivation could make use of external resources or just other attributes in need_info.

Technically a single callable could return a dict that gets merged, but I'm thinking it's better to be explicit about the values.

The downside here is this attribute only gets created at build time, but at least in our case we could have our CI builds could trigger the documentation build whenever it's updated. 

 I could also see a use case for a JavaScript method to get the value, but that would only be available to HTML builders and would probably be more complicated to implement.

**Comment by tlovett** [comment]
My only concern with this approach is that I have to specify that on every need vs. just declare it once on confy.py.  In my case, I want to apply this to all needs, but my thinking was if the callable returned None then the attribute would not be added to the need.  Can you expand on why you think that would be a bad thing?

I do see the utility in being to blend it in other attributes, but I don't think that fully replaces the other use case in my opinion.

**Comment by danwos** [comment]
Okay, I think this feature should be separated into 2 features:
* global_options and
* dynamic values

*Dynamic values* can be solved as mentioned above.
*Global_Options* can be defined via a variable in `conf.py` (as described above by you).

So the combination of both should fully support your case. 
But the benefit would be, that both features can be used separately and therefore supports in sum a lot of more use cases.

For *global options* I think a filter option would be great. So something like this:
 ```python
needs_global_options = [{
   'need_filter' = "'user story' == type and 'imported' in tags"
   'validated' : [[repos.is_validated]],
   'validation_link': [[repos.get_link]()],
   },
   {.. another global option ..}
]
```
All keys represent an option, which needs to be set on needs, which passed the filter.
But honestly the function of a dynamic value could also do the filtering by its own, so that there is no technical need to provide this directly in the definition.
But I think using a filter will be such a common function, that we should support it via definition.

**Comment by tlovett** [comment]
Yes, that makes sense to me.  

Should we split this into two separate issues then, and create a new one for dynamic values?

For the global options syntax, is there a reason you wrapped things in a list?  I was thinking the option would be set to a dict of attribute names to callables.  My thinking was that the callable would be responsible for the filtering, and simply return `None` if it wasn't applicable. That would avoid having to convert a string to Python code, and having the interplay of two different sets of logic being involved in setting the value.

Would that work for you?

**Comment by danwos** [comment]
Yes lets split it.
I have already renamed this issue to "global options" and created a new one with a rough description for dynamic values: #58.

The wrapped syntax was a mistake by me. 
For sure we don't need it in `conf.py`.
But something like this we will need in rst files, see #58.

We can skip the common filter option for the moment and let the called function do this step.
Will be easy to add such a filter thing later...

So for global options, this is how it should look, right?
```python
gloabl_options = {
   'validated' : "True",
   'validation_link':  "some_link",
}
```
And this will be set to ALL issues.
(How to use a function here is now part of #58 )

Do you agree?

**Comment by tlovett** [comment]
I'm good with this.  My only suggestion was to support the callables as part of this feature rather than #57.  Then all work for global options is done in one feature.

**Comment by danwos** [comment]
I'm fine with that.
But then we should implement #58 "Dynamic values" before this one.

**Comment by tlovett** [comment]
That works.  Thanks.

**Comment by danwos** [comment]
Implemented in 0.3.0
